Identify and check observations that can go to -109 C

jzuhone opened the pull request:
## Description

This PR adds the infrastructure necessary to check for -109 C violations on ACIS-S observations with the HETG or S3-only observations with < 300 expected total counts in the observing sequence. 

The bulk of this new functionality is the `fetch_ocat_data` function, which fetches data from the obscat using the web interface but requests that the table data be sent in text form. This allows us to query the obscat without using a username and password. It needs to obtain the information for each obsid, but then also check every obsid in the same sequence for a given obsid in the schedule, to ensure that the whole sequence has less than 300 expected counts if it is S3-only.

The filters for the various observation types are now put into a new function, `acis_filter`, which replaces `get_all_specific_instrument`. 

I also noted that in a few places the limit for obsid numbers was set incorrectly, so this has now been fixed. 

## Testing

- [x] Passes unit tests on MacOS
- [x] Functional testing

jzuhone committed:
A setup for catching "hot" ACIS-S observations

jzuhone committed:
Bug fixes

jzuhone committed:
Add viols test for the -109 limit

jzuhone committed:
Sort by time to match what we did before

jzuhone committed:
New test answers

jazan12 commented on acisfp_check/acis_obs.py:
```diff
@@ -101,7 +150,7 @@ def find_obsid_intervals(cmd_states):
 
         # Make sure we skip maneuver obsids explicitly
-        if 50000 > eachstate['obsid'] >= 38001:
+        if 60000 > eachstate['obsid'] >= 38001:
```
Do we want the 60000s excluded too?

jzuhone replied:
No because we need to check the cold ECS measurements as well.

jazan12 commented on acisfp_check/acis_obs.py:
```diff
@@ -40,11 +39,61 @@ def who_in_fp(simpos=80655):
+    # We figure out the CCD count from the table by finding out
+    # which ccds were on or optional, and then subtracting off
+    # the dropped chip count
+    ccd_count = np.zeros(tab["S3"].size, dtype='int')
```
Maybe misreading things but it looks to me like ccd_count is an array with length of your N_obsid, but a few lines down it appears you don't index into the array in adding / subtracting to ccd_count

jzuhone replied:
You are correct about the shape of the ccd_count array. All of the operations done to this array in the following lines are vectorized themselves, so we don't need to do indexing.

jazan12 reviewed (commented):
I followed along and had a couple questions I embedded as comments.  If you tell me those all check out ok then I'm happy to approve.

jzuhone committed:
Fix this for cold ecs

jzuhone committed:
We can simplify this by including cold ecs here

jzuhone committed:
Add cold ecs test

jzuhone committed:
Add a comment here

jzuhone commented:
I noticed that there wasn't a test for checking violations of cold ECS measurements, so I added one (and in the process found some bugs, for which I added new code). I also added some more comments.

jzuhone commented:
Note also that the answers changed for one of the violation tests. This occurred because if we had been using the -109 rules for previous loads, some observations which would have been limited to -111 C are no longer because they were HETG. In this case the changes in the answers are completely correct. Yay for good tests!

Gregg140 reviewed (approved):
Looks good to me.

jzuhone commented:
@taldcroft, @jeanconn, let me know if you'd like to have a look at this. In particular, have a look at the way the obscat was accessed.

jeanconn commented:
Thanks @jzuhone!  I have looked this over with regard to the ocat stuff. I was under the impression (from old email) that ACIS already had a history of using ocatDetails like https://cda.harvard.edu/srservices/ocatDetails.do?format=text&instrument=ACIS-I,ACIS- as part of load review.  Not sure if that was old information, but at least suggested to me that doing this kind of fetch did not add a new dependence (though if it does, I assume that needs to be documented someplace).

I've been looking at making a Ska data product with similar information from ocat via Sybase, but the ocatDetails fetch is nice text data.  

I also see from the SOT MP request this week about changing the Target Summary that they are putting data into the ocat that I assume could be used to do the same check?  But there is perhaps value in ACIS independently determining -109-ok observations from source data.

Gregg140 commented:
> Thanks @jzuhone! I have looked this over with regard to the ocat stuff. I was under the impression (from old email) that ACIS already had a history of using ocatDetails like https://cda.harvard.edu/srservices/ocatDetails.do?format=text&instrument=ACIS-I,ACIS- as part of load review. Not sure if that was old information, but at least suggested to me that doing this kind of fetch did not add a new dependence (though if it does, I assume that needs to be documented someplace).
> 
> I've been looking at making a Ska data product with similar information from ocat via Sybase, but the ocatDetails fetch is nice text data.
> 
> I also see from the SOT MP request this week about changing the Target Summary that they are putting data into the ocat that I assume could be used to do the same check? But there is perhaps value in ACIS independently determining -109-ok observations from source data.


A separate program does OCAT fetches.  However this is a better way to do it and I may change the older program to use this method.

jzuhone commented:
Hi @jeanconn, 

> Not sure if that was old information, but at least suggested to me that doing this kind of fetch did not add a new dependence (though if it does, I assume that needs to be documented someplace).

Do you think it's sufficient if it goes in our `acis_thermal_check` documentation, or does it need to be documented with Skare somewhere? 

> I also see from the SOT MP request this week about changing the Target Summary that they are putting data into the ocat that I assume could be used to do the same check? But there is perhaps value in ACIS independently determining -109-ok observations from source data.

It is valuable, but also this frees us from having to find and parse an OR list, which I think would involve more code and would probably be more error-prone.

jzuhone commented:
Now that I think about this more, I wonder what we would do if https://cda.harvard.edu was down?

jeanconn commented:
Right, so I thought you might already have a requirement on DS services and cda.harvard.edu being up to do load review.  If you need it to be more robust and didn't want to use the SOT MP product, it seems like you could:
1. set the code to fall back to emitting a warning and then do review not allowing any observations in that temperature range
1. wait for a Ska product with the pieces you want
1. make an ACIS product at time of OR list review (instead of load review) then used for review (or used as fallback)

jzuhone commented:
> Right, so I thought you might already have a requirement on DS services and cda.harvard.edu being up to do load review.

Yes, this is correct. But it would be nice for the program not to crash if it's not available for some reason. 

> If you need it to be more robust and didn't want to use the SOT MP product, it seems like you could:
> 1. set the code to fall back to emitting a warning and then do review not allowing any observations in that temperature range

I wrote some code to do this and I'm testing it presently

jzuhone commented:
> But it would be nice for the program not to crash if it's not available for some reason.

And I should have elaborated--by this I meant if you were running the program standalone for testing

jeanconn commented:
Gotcha.  Though a question then if just quitting cleanly with a handled
exception is fine for that.

On Tue, Sep 21, 2021, 3:27 PM John ZuHone ***@***.***> wrote:

> But it would be nice for the program not to crash if it's not available
> for some reason.
>
> And I should have elaborated--by this I meant if you were running the
> program standalone for testing
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/acisops/acisfp_check/pull/34#issuecomment-924316343>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGBHVEGYIQB55SKZHCWNQDUDDMA7ANCNFSM5EGFMSGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jzuhone committed:
If we can't access the obscat, then skip trying to check for -109 data

jzuhone committed:
If we're running this check after the observation has occurred, we need to check for the "D" entry in the table

taldcroft commented on acisfp_check/acis_obs.py:
```diff
@@ -40,11 +40,83 @@ def who_in_fp(simpos=80655):
+    params = {"obsid": obsid_list}
+    # First fetch the information from the obsid itself
+    got_table = True
+    try:
```
Just a heads-up that we have a handy function to allow robustly retrying a function (e.g. a web query, file access) that might momentarily fail but succeed a few seconds later.
https://github.com/sot/ska_helpers/blob/0867570c7dd1b44d22c1b61b8b3b44875a47c3b1/ska_helpers/retry/api.py#L108
Maybe overkill or not useful in this context, but we've found this useful especially in accessing HDF5 files on the NetApp.

jzuhone replied:
Hi @taldcroft, do you have a good code example where this is used?

jeanconn replied:
I know of a couple examples at https://github.com/sot/kadi/blob/593d705c023a2d982d30fe1cb74ba44b4d6006e2/kadi/commands/commands.py#L46 and https://github.com/sot/kadi/blob/593d705c023a2d982d30fe1cb74ba44b4d6006e2/kadi/update_cmds.py#L328

jzuhone replied:
This was a good suggestion--I've implemented it now.

jzuhone committed:
Implement retry_call

jzuhone commented:
@taldcroft @jeanconn let me know if you have any other comments or would like to approve.

jeanconn reviewed (approved):
Looks fine to me.
